Split out end points

benjefferies opened the pull request:
### What is the context of this PR?

Describe what you have changed and why, link to other PRs or Issues as appropriate.
- Split questionnaire submit and get requests
- Pull out thank-you page
- Pull out previous
- Remove first end point
  - first end point only ever appeared to be used in integration test
- Delete unused methods
### How to review

Describe the steps required to test the changes (include screenshots if appropriate).
- [x]  Run tests

benjefferies committed:
Split out end points

* Split questionnaire submit and get requests
* Pull out thank-you page
* Pull out previous
* Remove first end point
  * first end point only ever appeared to be used in integration test
* Delete unused methods

iwootten commented on app/main/views/questionnaire.py:
```diff
@@ -22,79 +18,91 @@
+@main_blueprint.route('/questionnaire/<eq_id>/<form_type>/<period_id>/<collection_id>/<location>', methods=["POST"])
+@login_required
+def post_questionnaire(eq_id, form_type, period_id, collection_id, location):
+    logger.debug("Post location : /questionnaire/%s/%s/%s/%s/%s", eq_id, form_type, period_id, collection_id, location)
```
Isn't flask (werkzeug) logging the url visited by default?

benjefferies replied:
From the look of my logs yes. Will remove them.

benjefferies committed:
Using decorator for state checks

iwootten commented on app/main/views/questionnaire.py:
```diff
@@ -22,79 +18,90 @@
-    except Exception as e:
-        return internal_server_error(e)
+    page = render_page('thank-you', questionnaire_manager)
+    questionnaire_manager.delete_user_data()
```
I feel that it may be good to move the questionnaire manager behaviour behind delete_user_data into this main route? It's only ever called here.

benjefferies replied:
Certainly, didn't catch that it was only used here.

benjefferies committed:
Remove delete user data function from 'qstore'

The delete_user_data function was only used by the questionnaire end
point and not in the questionnaire_store therefore refactoring to the
relevant place.

iwootten commented on app/main/views/questionnaire.py:
```diff
@@ -79,10 +80,15 @@ def get_thank_you(eq_id, form_type, period_id, collection_id):
 
+def delete_user_data():
+    get_questionnaire_store(current_user.user_id, current_user.user_ik).delete()
+    session_manager.clear()
```
Think you've missed a line here- need to call delete on the questionnaire store once it's retrieved.

iwootten replied:
Oh I see!

iwootten commented:
I think this change highlights how much questionnaire_manager is trying to do - for me I feel that some of that possibly should be tackled as part of this work?

benjefferies committed:
Splitting get_rendering_context method

* Splitting get_rendering_context logic into two methods
  * get_summary_rendering_context which loads the context for the
    summary page
  * get_rendering_context which loads contexts for all other pages
* Removing unused methods or moving methods which are used in one place
other than the questionnaire manager to that place:
  * get_current_state methods
  * _previous
  * submitted

benjefferies committed:
Merge remote-tracking branch 'origin/master' into refactor-endpoints

benjefferies committed:
Use common url pattern using blueprint prefix

* Using Blueprint(url_prefix=) to define common url prefix for
questionnaire end point.
* Splitting main_blueprint to questionnaire_blueprint and root_blueprint

benjefferies committed:
Correct logging parameters when processing answers

iwootten commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -345,29 +324,32 @@ def process_incoming_answers(self, location, post_data, replay=False):
             raise InvalidLocationException()
 
     def get_rendering_context(self, location):
```
If we're going to be splitting these into two methods, it probably makes sense that if the location is unknown within a call to get_rendering_context it should be capable of handling a location which is 'summary'. Maybe it should therefore check on the first line and return get_summary_rendering_context() if so?

benjefferies replied:
done

benjefferies committed:
Using flask error framework

* Making use of flasks error handling framework to take care of
  exception handling versus calling the error handling methods
  explicitly
* Writing integration test to check root blueprint exceptions are
  handled
* get_rendering_context handles summary location

benjefferies commented on tests/integration/mci/test_change_answers.py:
```diff
@@ -77,8 +77,3 @@ def happy_path(self, form_type_id, eq_id):
         self.assertRegexpMatches(content, '>Your responses<')
         self.assertRegexpMatches(content, 'Please check carefully before submission')
         self.assertRegexpMatches(content, '>Submit answers<')
-
```
The first end point is only ever used in integration tests so the end point and related tests have been removed

benjefferies committed:
Merge remote-tracking branch 'origin/master' into refactor-endpoints

benjefferies commented on tests/app/views/test_health_check.py:
```diff
@@ -2,9 +2,3 @@ def test_healthcheck(survey_runner):
     assert "success" in response.get_data(as_text=True)
-
-
-def test_errors_404(survey_runner):
```
This exact test live in test_errors.py

ajmaddaford commented on app/main/views/questionnaire.py:
```diff
@@ -1,103 +1,148 @@
+    return get_page(questionnaire_manager, location, is_valid)
+
+
+def get_page(questionnaire_manager, location, is_valid):
```
get_page is confusing, I think render_page would be clearer

benjefferies replied:
Good call

ajmaddaford commented on app/main/views/questionnaire.py:
```diff
@@ -1,103 +1,148 @@
-    else:
-        return do_get(questionnaire_manager, next_location, is_valid)
+def get_rendering_template(location):
+    return TemplateRegistry.get_template_name(location)
```
Any reason not to use this directly rather than via the `get_rendering_template` method?

benjefferies replied:
Nope, copy and paste initially and then forgot to remove

ajmaddaford commented on app/main/views/questionnaire.py:
```diff
@@ -1,103 +1,148 @@
-        metadata = get_metadata(current_user)
-        current_survey = eq_id + form_type + period_id + collection_id
-        metadata_survey = metadata["eq_id"] + metadata["form_type"] + metadata["period_id"] + metadata["collection_exercise_sid"]
+    return decorated_function
```
I think we should re-implement this using the `url_value_preprocessor` blueprint decorator function (see http://exploreflask.com/en/latest/blueprints.html#using-a-dynamic-url-prefix). We should also set the questionnaire_manager on the global application (request) context so that the routes that use this function don't need to re-instantiate the manager.

benjefferies replied:
Done

ajmaddaford commented on app/main/views/questionnaire.py:
```diff
@@ -1,103 +1,148 @@
 
 
 def do_redirect(eq_id, form_type, period_id, collection_id,  location):
     return redirect('/questionnaire/' + eq_id + '/' + form_type + '/' + period_id + '/' + collection_id + '/' + location)
```
We should be using the Flask url_for method for generating URLs for known routes. Also, is there a way to get the current route rather than keep having to pass all of the parameters to this method?

benjefferies replied:
Done

ajmaddaford commented on app/main/views/questionnaire.py:
```diff
@@ -1,103 +1,148 @@
+    return current_survey == metadata_survey
+
+
+def go_to_page(collection_id, eq_id, form_type, period_id, location, is_valid=True):
```
This method go's to a page **and** renders it. I think we should either rename the method or move the rendering out to be called independently.

benjefferies replied:
Not a great fan of methods with `and` suggests it's doing too much which could be right in this case so I've pulled out

ajmaddaford commented on app/main/views/errors.py:
```diff
@@ -13,47 +17,39 @@
 
 logger = logging.getLogger(__name__)
 
+errors_blueprint = Blueprint('errors', __name__)
```
Should the errors blueprint be setting the cache headers like the root and questionnaire blueprints?

benjefferies replied:
What is the reason we are indicating to the browser not to cache? I can add it just to blanket cover this too anyway

benjefferies committed:
Refactor questionnaire logic

* Using flask.before_request to check questionnaire state
* Removing unnecessary methods
* Renaming methods
* Creating catch all for InvalidLocationException using flask framework
* Removing method only used in tests

benjefferies commented:
Superseded by https://github.com/ONSdigital/eq-survey-runner/pull/512
